hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java

```diff
@@ -0,0 +1,553 @@
+import static com.hazelcast.jet.core.processor.SourceProcessors.convenientTimestampedSourceP;
+
+/**
+ * Top-level class for Jet source builders. Refer to the builder
+ * factory methods listed below.
+ *
+ * @see #batch(String, DistributedFunction)
+ * @see #timestampedStream(String, DistributedFunction)
+ * @see #stream(String, DistributedFunction)
+ *
+ * @param <S> type of the source state object
+ */
+public final class SourceBuilder<S> {
+    // the m-prefix in field names is there to allow unqualified access from
+    // inner classes while not clashing with parameter names. Parameter names
+    // are public API.
+    private final String mName;
+    private final DistributedFunction<? super Processor.Context, ? extends S> mCreateFn;
+    private DistributedConsumer<? super S> mDestroyFn;
+    private int mPreferredLocalParallelism;
+
+    /**
+     * The buffer object that the {@code fillBufferFn} gets on each call. Used
+     * in sources that emit items without a timestamp.
+     *
+     * @param <T> type of the buffer item
+     */
+    public interface SourceBuffer<T> {
+
+        /**
+         * Adds an item to the buffer.
+         */
+        void add(T item);
+
+        /**
+         * Closes the buffer, signaling that all items have been emitted.
+         */
+        void close();
+    }
+
+    /**
+     * The buffer object that the {@code fillBufferFn} gets on each call. Used
+     * in sources that emit timestamped items.
+     *
+     * @param <T> type of the buffer item
+     */
+    public interface TimestampedSourceBuffer<T> {
+
+        /**
+         * Adds an item to the buffer, assigning a timestamp to it. The timestamp
+         * is in milliseconds.
+         */
+        void add(T item, long timestamp);
+
+        /**
+         * Closes the buffer, signaling that all items have been emitted.
+         */
+        void close();
+    }
+
+    private SourceBuilder(
+            @Nonnull String name,
+            @Nonnull DistributedFunction<? super Processor.Context, ? extends S> createFn
+    ) {
+        this.mName = name;
+        this.mCreateFn = createFn;
+    }
+
+    /**
+     * Returns a fluent-API builder with which you can create a {@linkplain
+     * BatchSource batch source} for a Jet pipeline.
+     * <p>
+     * Each parallel worker that drives your source has its private instance of
+     * a state object it gets from your {@code createFn}. To get the data items
+     * to emit to the pipeline, the worker repeatedly calls your {@code
+     * fillBufferFn} with the state object and a buffer object.
+     * <p>
+     * Your function should add some items to the buffer, ideally those it has
+     * ready without having to block. It shouldn't add more than a thousand
+     * items at once, but there is no limit enforced. If it doesn't have any
+     * items ready, it may also return without adding anything. In any case the
+     * function should not take more than a second or so to complete, otherwise
```

Comment:
It would be good to say that "the processor is non-cooperative, it is allowed to do blocking calls"